core/unix/os.c

```diff
@@ -788,6 +788,82 @@ static init_fn_t
 #else
 /* If we're a normal shared object, then we override _init.
  */
+
+#    if defined(MUSL)
+#        define EFAULT 14
+#        define AT_EXECFN 31
+static int
+check_address_readable(void *addr)
+{
+    return dynamorio_syscall(SYS_rt_sigprocmask, ~0L, addr, NULL,
+                             sizeof(kernel_sigset_t)) != EFAULT;
+}
+
+/* When entering the entry point, the stack layout looks like
+ *   sp => argc
+ *         argv
+ *         NULL (end of argv)
+ *         envp
+ *         NULL (end of envp)
+ *         auxv
+ * search_auxvector() walks towards the higher address and locate one of the
+ * auxvector entry, then walk backwards and find the beginning of auxvector. */
+static void *
+search_auxvector(void *sp)
+{
+    /* XXX: Check whether 64 * PAGE_SIZE is an appropriate limit */
+    for (size_t offset = 0; offset < PAGE_SIZE * 64; offset += sizeof(ulong)) {
+        ELF_AUXV_TYPE *p = sp + offset;
+
+        if (((uintptr_t)(&p->a_un) & (PAGE_SIZE - 1)) == 0 &&
+            !check_address_readable(&p->a_un))
+            return NULL;
+
+        /* Check for AT_EXECFN entry in the auxvector, which contains pathname
+         * of the program and should be a readable address. */
+        if (p->a_type == AT_EXECFN && check_address_readable((void *)p->a_un.a_val)) {
```

Comment:
Is this a strong enough check?  An integer 31 adjacent to a valid pointer?  It seems like this could happen to appear somewhere else on the stack.

Reply:
Since limited to ifdef MUSL that may be ok.  Make sure to add a comment about it being new.  And a comment about the fragility of the fallback if it's left in place.